torch/distributions/constraints.py

```diff
@@ -102,7 +122,49 @@ def __init__(self, low, high):
             def support(self):
                 return constraints.interval(self.low, self.high)
     """
-    pass
+    def __init__(self, fn=None, *, is_discrete=False, event_dim=0):
+        self.is_discrete = is_discrete
+        self.event_dim = event_dim
+        super().__init__(fn)
+
+    def __call__(self, fn):
+        """
+        Support for syntax to customize static attributes::
+
+            @constraints.dependent_property(is_discrete=True, event_dim=1)
+            def support(self):
+                ...
+        """
+        return _DependentProperty(fn, is_discrete=self.is_discrete, event_dim=self.event_dim)
+
+
+class _IndependentConstraint(Constraint):
+    """
+    Wraps a constraint by aggregating over ``reinterpreted_batch_ndims``-many
+    dims in :meth:`check`, so that an event is valid only if all its
+    independent entries are valid.
+    """
+    def __init__(self, base_constraint, reinterpreted_batch_ndims):
+        assert isinstance(base_constraint, Constraint)
+        assert isinstance(reinterpreted_batch_ndims, int)
+        assert reinterpreted_batch_ndims >= 0
+        self.base_constraint = base_constraint
+        self.reinterpreted_batch_ndims = reinterpreted_batch_ndims
+        super().__init__()
+
+    @property
+    def is_discrete(self):
+        return self.base_constraint.is_discrete
+
+    @property
+    def event_dim(self):
+        return self.base_constraint.event_dim + self.reinterpreted_batch_ndims
+
+    def check(self, value):
+        result = self.base_constraint.check(value)
```

Comment:
Should we also raise an error here if the value's ndim < event_dim ? e.g. something like `independent(constraints.positive, 2).check(torch.ones(10))` should probably throw an error. WDYT?

Reply:
done